sqlite.go

```diff
@@ -275,6 +275,32 @@ func (c *conn) BeginTx(ctx context.Context, opts driver.TxOptions) (driver.Tx, e
 // Raw is so ConnInitFunc can cast to SQLConn.
 func (c *conn) Raw(fn func(any) error) error { return fn(c) }
 
+// ResetSession is called prior to executing a query on the connection
+// if the connection has been used before. If the driver returns ErrBadConn
+// the connection is discarded.
+func (c *conn) ResetSession(ctx context.Context) error {
+	if c.txState != txStateNone {
+		return errors.New("cannot ResetSession on a connection with an active transaction")
+	}
+
+	// ensure that all prepared statements are reset and unbound
+	if c.stmts != nil {
+		for _, s := range c.stmts {
+			if err := s.resetAndClear(); err != nil {
+				return err
```

Comment:
Should this return ErrBadConn?

Reply:
In the above case we could and probably should. In this particular case, that's unclear because:

```
ErrBadConn should be returned by a driver to signal to the sql package that a driver.Conn is in a bad state (such as the server having earlier closed the connection) and the sql package should retry on a new connection.

To prevent duplicate operations, ErrBadConn should NOT be returned if there's a possibility that the database server might have performed the operation. Even if the server sends back an error, you shouldn't return ErrBadConn.

Errors will be checked using errors.Is. An error may wrap ErrBadConn or implement the Is(error) bool method.
```

If we have potentially completed a tx we can not disambiguate between these conditions so we should not match BadConn.